### ISSUE TITLE
Expose CompilerDaemon functionality to Plugin Authors

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#1239]
When implementing compilation support for a JVM language, it is desirable to leverage the performance advantages of daemonized compiler workers - more commonly known as forking compilation. A forked compiler process is usually persisted and reused across multiple projects in a multi-project build. While forked compilation is possible with Java, Groovy and Scala, classloader isolation prevents a 3rd party plugin from delegating work to a daemonized worker process. 

[Design proposal from Kyle Moore](https://docs.google.com/document/d/1JqwTn0LV1sw-CHB8r4AUML0d-YLlQ6GNUYsGjYQal1Y/edit).

### Expected Behavior
As a plugin-developer attempting to add support for a new language, I want to be able to run my compiler in a forked (and possibly long-running) process which can survive from build to build, so that I can take advantage of my language's compiler features like incremental compilation.

### Current Behavior
We do not have a public API for plugins to schedule work in a compiler daemon.

### Context
Guidewire would like to have support for their GoSu compiler. They even put together this awesome pull request to add it to the Gradle distribution. Unfortunately, we cannot take on the engineering complexity or distribution bloat to support GoSu due to its relatively small user-base. 

JetBrains is interested in adding Kotlin support.

Grails probably needs the same feature for the "unsafe incremental compiler"

**Comment by pioterj** [comment]
@oehme This will be covered by the work on worker processes, right? Can we set 4.0 milestone on it?

**Comment by oehme** [comment]
That's right, we'll add some documentation by then.

**Comment by eriwen** [comment]
@oehme assigning to you to close this once documentation is done.

**Comment by oehme** [comment]
We've decided to postpone user guide and blog post until we use the API in our own code base (e.g. for Java compilation) in 4.1. Nevertheless, the API is already functional and has some rudimentary JavaDoc, so I encourage interested plugin authors to [give it a first look.](https://github.com/gradle/gradle/blob/master/subprojects/workers/src/main/java/org/gradle/workers/WorkerExecutor.java)

**Comment by oehme** [comment]
@ghale can you close this once #2479 is merged?

**Comment by ghale** [comment]
#2479 Has been merged.
